File: crates/dashboard-kit/web/src/posture/claims.ts
import type { CapabilityStatus, ProtectionLayer, ScopeRef, VersionRef } from "../api/v1";

export type ContainmentClaimContext = {
  matrix: VersionRef;
  claim_id: string;
  scope_id: string;
  scope_kind: ScopeRef["kind"];
  action_class: string;
  population: string;
  environment: string;
  /** Producer snapshot generation time. */
  generated_at: string;
  /** Consumer-controlled clock used to expire frozen snapshots. */
  evaluated_at: string;
};

const verifiedEvidence = (evidence: { integrity: string }[]) =>
  evidence.length > 0 && evidence.every((item) => item.integrity === "verified");

const hasHostVerifiedScope = (scope: ScopeRef[], context: ContainmentClaimContext) =>
  scope.some((item) => item.id === context.scope_id
    && item.kind === context.scope_kind
    && item.verification === "host_verified"
    && verifiedEvidence(item.evidence));

const matrixIsPinned = (matrix: VersionRef) => matrix.id.trim() !== ""
  && matrix.version.trim() !== ""
  && /^sha256:[a-f0-9]{64}$/.test(matrix.digest);

const sameMatrix = (left: VersionRef, right: VersionRef) => left.id === right.id
  && left.version === right.version
  && left.canonicalization === right.canonicalization
  && left.digest === right.digest;

const freshnessIsCurrent = (
  freshness: CapabilityStatus["freshness"],
  context: ContainmentClaimContext,
) => {
  if (freshness.state !== "fresh" || freshness.budget_seconds <= 0
    || freshness.observed_at === null || freshness.age_seconds === null) return false;
  const generatedMs = Date.parse(context.generated_at);
  const evaluatedMs = Date.parse(context.evaluated_at);
  const observedMs = Date.parse(freshness.observed_at);
  if (![generatedMs, evaluatedMs, observedMs].every(Number.isFinite)) return false;
  if (observedMs > generatedMs || generatedMs > evaluatedMs) return false;
  const producerAge = Math.floor((generatedMs - observedMs) / 1_000);
  const consumerAge = Math.floor((evaluatedMs - observedMs) / 1_000);
  return producerAge >= 0
    && consumerAge >= 0
    && freshness.age_seconds === producerAge
    && consumerAge <= freshness.budget_seconds;
};

const evidenceMatchesFreshness = (
  evidence: CapabilityStatus["claims"][number]["evidence"],
  freshness: CapabilityStatus["freshness"],
) => verifiedEvidence(evidence)
  && evidence.some((item) => item.observed_at === freshness.observed_at
    && item.freshness.observed_at === freshness.observed_at
    && item.freshness.budget_seconds === freshness.budget_seconds
    && item.freshness.state === freshness.state
    && item.freshness.age_seconds === freshness.age_seconds);

const verifiedStage = (
  stage: CapabilityStatus["convergence"]["configured"],
  freshness: CapabilityStatus["freshness"],
) => stage.state === "yes" && evidenceMatchesFreshness(stage.evidence, freshness);

const contextIsBound = (context: ContainmentClaimContext) => matrixIsPinned(context.matrix)
  && context.claim_id.trim() !== ""
  && context.scope_id.trim() !== ""
  && context.action_class.trim() !== ""
  && context.population.trim() !== ""
  && context.environment.trim() !== "";

const claimSupportsContext = (
  claim: CapabilityStatus["claims"][number],
  context: ContainmentClaimContext,
  freshness: CapabilityStatus["freshness"],
) => claim.id === context.claim_id
  && claim.status === "verified"
  && claim.versions.some((version) => sameMatrix(version, context.matrix))
  && ((claim.statement?.trim().length ?? 0) > 0 || (claim.semantic_key?.trim().length ?? 0) > 0)
  && claim.population === context.population
  && claim.environment === context.environment
  && claim.limitations.length === 0
  && claim.observed_at === freshness.observed_at
  && claim.reviewed_at !== null
  && claim.expires_at !== null
  && Number.isFinite(Date.parse(claim.reviewed_at))
  && Number.isFinite(Date.parse(claim.expires_at))
  && Date.parse(claim.reviewed_at) <= Date.parse(claim.expires_at)
  && Date.parse(context.evaluated_at) <= Date.parse(claim.expires_at)
  && hasHostVerifiedScope(claim.scope, context)
  && claim.action_classes.includes(context.action_class)
  && evidenceMatchesFreshness(claim.evidence, freshness);

/**
 * Presentation guardrail only: the backend remains the evidence authority. This
 * makes the browser fail closed if unsupported, declared, stale or weakly
 * evidenced data is accidentally labelled as active host containment.
 */
export function capabilityMayClaimActiveContainment(
  capability: CapabilityStatus,
  context: ContainmentClaimContext,
): boolean {
  return capability.tier === "enterprise_core"
    && capability.availability === "available"
    && capability.support === "supported"
    && capability.effective_mode === "enforce"
    && capability.rollout_state === "enforcing"
    && capability.health === "healthy"
    && contextIsBound(context)
    && freshnessIsCurrent(capability.freshness, context)
    && verifiedStage(capability.convergence.configured, capability.freshness)
    && verifiedStage(capability.convergence.loaded, capability.freshness)
    && verifiedStage(capability.convergence.running, capability.freshness)
    && verifiedStage(capability.convergence.enforcing, capability.freshness)
    && verifiedStage(capability.convergence.verified_effective, capability.freshness)
    && hasHostVerifiedScope(capability.scope, context)
    && capability.covered_action_classes.includes(context.action_class)
    && capability.bypass_classes.length === 0
    && capability.known_uncovered_paths.length === 0
    && capability.last_evidence?.integrity === "verified"
    && evidenceMatchesFreshness([capability.last_evidence], capability.freshness)
    && capability.claims.some((claim) => claimSupportsContext(
      claim,
      context,
      capability.freshness,
    ));
}

export function layerMayClaimActiveContainment(
  layer: ProtectionLayer,
  capability: CapabilityStatus,
  layerContext: ContainmentClaimContext,
  capabilityContext: ContainmentClaimContext,
): boolean {
  return layer.capability_ids.includes(capability.id)
    && sameClaimTarget(layerContext, capabilityContext)
    && capabilityMayClaimActiveContainment(capability, capabilityContext)
    && layer.claim_state === "active"
    && layer.effective_mode === "enforce"
    && contextIsBound(layerContext)
    && freshnessIsCurrent(layer.freshness, layerContext)
    && layer.known_gaps.length === 0
    && hasHostVerifiedScope(layer.effective_scope, layerContext)
    && layer.covered_action_classes.includes(layerContext.action_class)
    && evidenceMatchesFreshness(layer.evidence, layer.freshness)
    && verifiedStage(layer.convergence.configured, layer.freshness)
    && verifiedStage(layer.convergence.loaded, layer.freshness)
    && verifiedStage(layer.convergence.running, layer.freshness)
    && verifiedStage(layer.convergence.enforcing, layer.freshness)
    && verifiedStage(layer.convergence.verified_effective, layer.freshness);
}

const sameClaimTarget = (
  left: ContainmentClaimContext,
  right: ContainmentClaimContext,
) => sameMatrix(left.matrix, right.matrix)
  && left.claim_id === right.claim_id
  && left.scope_id === right.scope_id
  && left.scope_kind === right.scope_kind
  && left.action_class === right.action_class
  && left.population === right.population
  && left.environment === right.environment
  && left.evaluated_at === right.evaluated_at;

File: crates/dashboard-kit/src/claims.rs
//! Presentation-side safety rails for protection and containment wording.
//!
//! These helpers cannot prove a control effective; only the producer can emit
//! runtime evidence. They fail closed when that evidence is missing, stale,
//! unhealthy, out of scope, or not converged.

use crate::contract::{
    Availability, CapabilityStatus, CapabilityTier, ClaimState, ClaimStatus, EffectiveMode,
    EvidenceFreshness, EvidenceRef, FreshnessState, HealthState, IntegrityState, ProtectionLayer,
    ScopeKind, ScopeRef, ScopeVerification, StageAnswer, StageState, SupportLevel, VersionRef,
};

/// The exact claim the UI intends to render. A global "protected" boolean is
/// deliberately not supported: every affirmative statement must be bound to a
/// reviewed matrix revision, host-verified scope, action class and response
/// generation time.
pub struct ContainmentClaimContext<'a> {
    pub matrix: &'a VersionRef,
    pub claim_id: &'a str,
    pub scope_id: &'a str,
    pub scope_kind: ScopeKind,
    pub action_class: &'a str,
    pub population: &'a str,
    pub environment: &'a str,
    /// Time at which the producer assembled the response. This is checked for
    /// internal age coherence, but is never treated as the current clock.
    pub generated_at: &'a str,
    /// Consumer-controlled evaluation time. Injecting this makes expiry
    /// deterministic in tests and prevents a frozen producer snapshot from
    /// refreshing its own freshness forever.
    pub evaluated_at: &'a str,
}

pub fn capability_may_claim_active_containment(
    capability: &CapabilityStatus,
    context: &ContainmentClaimContext<'_>,
) -> bool {
    capability.availability == Availability::Available
        && capability.tier == CapabilityTier::EnterpriseCore
        && capability.support == SupportLevel::Supported
        && capability.effective_mode == EffectiveMode::Enforce
        && capability.rollout_state == crate::contract::RolloutState::Enforcing
        && capability.health == HealthState::Healthy
        && matrix_is_pinned(context.matrix)
        && freshness_is_current(&capability.freshness, context)
        && stage_is_verified(&capability.convergence.configured, &capability.freshness)
        && stage_is_verified(&capability.convergence.loaded, &capability.freshness)
        && stage_is_verified(&capability.convergence.running, &capability.freshness)
        && stage_is_verified(&capability.convergence.enforcing, &capability.freshness)
        && stage_is_verified(
            &capability.convergence.verified_effective,
            &capability.freshness,
        )
        && scope_is_host_verified(
            &capability.scope,
            context.scope_id,
            context.scope_kind,
        )
        && capability
            .covered_action_classes
            .iter()
            .any(|action| action == context.action_class)
        // The foundation DTO cannot yet bind these free-form gaps to a precise
        // scope/action pair. Fail closed until the Assurance Matrix adapter has
        // proven they do not apply to this requested statement.
        && capability.bypass_classes.is_empty()
        && capability.known_uncovered_paths.is_empty()
        && capability.last_evidence.as_ref().is_some_and(|evidence| {
            evidence_matches_freshness(evidence, &capability.freshness)
        })
        && capability.claims.iter().any(|claim| {
            claim_supports_context(claim, context, &capability.freshness)
        })
}

pub fn layer_may_claim_active_containment(
    layer: &ProtectionLayer,
    capability: &CapabilityStatus,
    layer_context: &ContainmentClaimContext<'_>,
    capability_context: &ContainmentClaimContext<'_>,
) -> bool {
    layer.capability_ids.iter().any(|id| id == &capability.id)
        && same_claim_target(layer_context, capability_context)
        && capability_may_claim_active_containment(capability, capability_context)
        && layer.claim_state == ClaimState::Active
        && layer.effective_mode == EffectiveMode::Enforce
        && matrix_is_pinned(layer_context.matrix)
        && freshness_is_current(&layer.freshness, layer_context)
        && layer.known_gaps.is_empty()
        && layer
            .covered_action_classes
            .iter()
            .any(|action| action == layer_context.action_class)
        && scope_is_host_verified(
            &layer.effective_scope,
            layer_context.scope_id,
            layer_context.scope_kind,
        )
        && evidence_set_is_verified(&layer.evidence, &layer.freshness)
        && stage_is_verified(&layer.convergence.configured, &layer.freshness)
        && stage_is_verified(&layer.convergence.loaded, &layer.freshness)
        && stage_is_verified(&layer.convergence.running, &layer.freshness)
        && stage_is_verified(&layer.convergence.enforcing, &layer.freshness)
        && stage_is_verified(&layer.convergence.verified_effective, &layer.freshness)
}

fn same_claim_target(
    left: &ContainmentClaimContext<'_>,
    right: &ContainmentClaimContext<'_>,
) -> bool {
    left.matrix == right.matrix
        && left.claim_id == right.claim_id
        && left.scope_id == right.scope_id
        && left.scope_kind == right.scope_kind
        && left.action_class == right.action_class
        && left.population == right.population
        && left.environment == right.environment
        && left.evaluated_at == right.evaluated_at
}

fn matrix_is_pinned(matrix: &VersionRef) -> bool {
    !matrix.id.trim().is_empty()
        && !matrix.version.trim().is_empty()
        && matrix.digest.len() == 71
        && matrix.digest.starts_with("sha256:")
        && matrix.digest[7..]
            .bytes()
            .all(|byte| byte.is_ascii_digit() || (b'a'..=b'f').contains(&byte))
}

fn scope_is_host_verified(scopes: &[ScopeRef], id: &str, kind: ScopeKind) -> bool {
    scopes.iter().any(|scope| {
        scope.id == id
            && scope.kind == kind
            && scope.verification == ScopeVerification::HostVerified
            && !scope.evidence.is_empty()
            && scope
                .evidence
                .iter()
                .all(|evidence| evidence.integrity == IntegrityState::Verified)
    })
}

fn freshness_is_current(
    freshness: &EvidenceFreshness,
    context: &ContainmentClaimContext<'_>,
) -> bool {
    if freshness.state != FreshnessState::Fresh || freshness.budget_seconds == 0 {
        return false;
    }
    let (Some(observed_at), Some(reported_age)) =
        (freshness.observed_at.as_deref(), freshness.age_seconds)
    else {
        return false;
    };
    let Ok(generated_at) = chrono::DateTime::parse_from_rfc3339(context.generated_at) else {
        return false;
    };
    let Ok(evaluated_at) = chrono::DateTime::parse_from_rfc3339(context.evaluated_at) else {
        return false;
    };
    let Ok(observed_at) = chrono::DateTime::parse_from_rfc3339(observed_at) else {
        return false;
    };
    let producer_age = generated_at
        .signed_duration_since(observed_at)
        .num_seconds();
    let consumer_age = evaluated_at
        .signed_duration_since(observed_at)
        .num_seconds();
    observed_at <= generated_at
        && generated_at <= evaluated_at
        && producer_age >= 0
        && consumer_age >= 0
        && u64::try_from(producer_age).ok() == Some(reported_age)
        && u64::try_from(consumer_age)
            .ok()
            .is_some_and(|age| age <= freshness.budget_seconds)
}

fn claim_supports_context(
    claim: &crate::contract::ClaimRecord,
    context: &ContainmentClaimContext<'_>,
    freshness: &EvidenceFreshness,
) -> bool {
    let (Some(reviewed_at), Some(expires_at)) =
        (claim.reviewed_at.as_deref(), claim.expires_at.as_deref())
    else {
        return false;
    };
    let Ok(reviewed_at) = chrono::DateTime::parse_from_rfc3339(reviewed_at) else {
        return false;
    };
    let Ok(expires_at) = chrono::DateTime::parse_from_rfc3339(expires_at) else {
        return false;
    };
    let Ok(evaluated_at) = chrono::DateTime::parse_from_rfc3339(context.evaluated_at) else {
        return false;
    };

    claim.id == context.claim_id
        && claim.status == ClaimStatus::Verified
        && claim
            .versions
            .iter()
            .any(|version| version == context.matrix)
        && claim
            .statement
            .as_deref()
            .is_some_and(|statement| !statement.trim().is_empty())
        && claim.population == context.population
        && claim.environment == context.environment
        && claim.limitations.is_empty()
        && claim.observed_at.as_deref() == freshness.observed_at.as_deref()
        && reviewed_at <= expires_at
        && evaluated_at <= expires_at
        && scope_is_host_verified(&claim.scope, context.scope_id, context.scope_kind)
        && claim
            .action_classes
            .iter()
            .any(|action| action == context.action_class)
        && evidence_set_is_verified(&claim.evidence, freshness)
}

fn stage_is_verified(stage: &StageState, freshness: &EvidenceFreshness) -> bool {
    stage.state == StageAnswer::Yes && evidence_set_is_verified(&stage.evidence, freshness)
}

fn evidence_set_is_verified(evidence: &[EvidenceRef], freshness: &EvidenceFreshness) -> bool {
    !evidence.is_empty()
        && evidence
            .iter()
            .all(|item| item.integrity == IntegrityState::Verified)
        && evidence
            .iter()
            .any(|item| evidence_matches_freshness(item, freshness))
}

fn evidence_matches_freshness(evidence: &EvidenceRef, freshness: &EvidenceFreshness) -> bool {
    evidence.integrity == IntegrityState::Verified
        && evidence.freshness == *freshness
        && freshness.observed_at.as_deref() == Some(evidence.observed_at.as_str())
}
